mkosi/__init__.py

```diff
@@ -2218,21 +2218,61 @@ def finalize_initrds(context: Context) -> list[Path]:
     return [build_default_initrd(context)]
 
 
+def expand_kernel_specifiers(text: str, kver: str, token: str, roothash: str, boot_count: str) -> str:
+    specifiers = {
+        "&": "&",
+        "e": token,
+        "k": kver,
+        "h": roothash,
+        "c": boot_count
+    }
+
+    def replacer(match: re.Match[str]) -> str:
+        m = match.group("specifier")
+        if specifier := specifiers.get(m):
+            return specifier
+
+        logging.warning(f"Unknown specifier '&{m}' found in {text}, ignoring")
+        return ""
+
+    return re.sub(r"&(?P<specifier>[&a-zA-Z])", replacer, text)
+
+
 def install_type1(
     context: Context,
     kver: str,
     kimg: Path,
     token: str,
     partitions: Sequence[Partition],
 ) -> None:
-    dst = context.root / "boot" / token / kver
-    entry = context.root / f"boot/loader/entries/{token}-{kver}.conf"
+    roothash = ""
+    if roothash_cli := finalize_roothash(partitions):
+        roothash = roothash_cli.partition("=")[2]
+
+    boot_count = ""
+    if (context.root / "etc/kernel/tries").exists():
+        boot_count = (context.root / "etc/kernel/tries").read_text().strip()
+
+    bootloader_entry = context.config.bootloader_entry_format or "&e-&k"
```

Comment:
If this would be changed to a default value it wouldn't be possible append `roothash` and `tries` in the default for UKI without reintroducing the current issue of not being able to remove them.

If its the default value you shouldn't append to it, as to not change what the user specified and comparing if `context.config.bootloader_entry_format != "&e-&k"` would be problematic if the user specifically just wanted that format in the UKI case without `roothash` and `tries`.

Reply:
My thinking was having this code:
```python
    bootloader_entry = context.config.bootloader_entry_format
```
and setting `"&e-&k"` as the default in `config.py`, instead of the current `None`. `bootloader_entry` can still be changed locally, i.e. root hash and boot count can be appended, but the default of `"&e-&k"` is put in a central place. Am I missing something?

Reply:
If you'd set `&e-&k` as the default for `bootloader_entry_format` there isn't a way for the UKI case to figure out if it should try and append the roothash and boot count for the default case. [Currently it relies on if the value is `None` or not to figure that out.](https://github.com/NekkoDroid/mkosi/blob/8e2b6efb5459100d764d9101f78abec721b54abd/mkosi/__init__.py#L2356)

I would be fine dropping the roothash/tries logic and just using the value set, but since this would be a change in default behavior I'd need the OK on that if it really is something that is wanted.

Reply:
Ok, now I see it. Let's move out the `&e-&k` into some global variable `DEFAULT_BOOTLOADER_ENTRY` and add a comment why this is and in `config.py`, that the default is in this variable.